File: leetcode_cn/24.py
# %%
from typing import List
# Definition for singly-linked list.


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        dn = ListNode(-1, head)
        if head is None or head.next is None:
            return head
        h = dn
        while h and h.next and h.next.next:
            a = h
            b = h.next
            c = h.next.next
            a.next = c
            b.next = c.next
            c.next = b
            h = b
        return dn.next

File: lcof/24.cpp
#include <bits/stdc++.h>
using namespace std;
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
  public:
    ListNode *reverseList(ListNode *head) {
        ListNode *p1 = nullptr;
        ListNode *p2 = head;
        ListNode *p3;
        while (p2) {
            p3 = p2->next;
            p2->next = p1;
            p1 = p2;
            p2 = p3;
        }
        return p1;
    }
};
int main() {
    //
    return 0;
}